// nestjs-form-data/dist/decorators/validation/has-mime-type.validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HasMimeType = void 0;
const class_validator_1 = require("class-validator");
const is_file_validator_1 = require("./is-file.validator");
function HasMimeType(allowedMimeTypes, validationOptions) {
    return class_validator_1.ValidateBy({
        name: 'HasMimeType',
        constraints: [allowedMimeTypes],
        validator: {
            validate(value, args) {
                const allowedMimeTypes = args.constraints[0] || [];
                if (is_file_validator_1.isFile(value)) {
                    return allowedMimeTypes.includes(value.mimetype);
                }
                return false;
            },
            defaultMessage(validationArguments) {
                const allowedMimeTypes = validationArguments.constraints[0] || [];
                return `File must be of one of the types ${allowedMimeTypes.join(', ')}`;
            },
        },
    }, validationOptions);
}
exports.HasMimeType = HasMimeType;
